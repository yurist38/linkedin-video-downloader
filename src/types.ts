export enum Actions {
  AddButtons = 'addButtons',
  Download = 'download',
}

export interface Message {
  action: Actions;
  url?: string;
}

export interface Options {
  buttonPosition: {
    horizontal: 'left' | 'right';
    vertical: 'top' | 'bottom';
  };
}

export enum CommonNames {
  ButtonClassName = 'linkedin-video-downloader-btn',
  OptionsStorageKey = 'linkedin-video-downloader-options',
}
