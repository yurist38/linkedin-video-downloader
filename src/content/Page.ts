import { Message, Actions } from '../types';
import { debounce } from 'ts-debounce';

class Page {
  static buttonClassName = 'linkedin-video-downloader-btn';
  public buttonElement = Page.createDownloadButton();
  public observerCallbackDebounced = debounce(this.observerCallback, 1000);
  public observer = new MutationObserver(this.observerCallbackDebounced.bind(this));

  public constructor() {
    this.observer.observe(window.document, {
      childList: true,
      subtree: true,
    });
    chrome.runtime.onMessage.addListener(this.onMessage.bind(this));
  }

  public observerCallback(): void {
    this.processAllVideos();
  }


  static createDownloadButton(): HTMLElement {
    const button = document.createElement('button');
    const img = document.createElement('img');
    button.classList.add(Page.buttonClassName);
    button.title = 'Download this video!';
    button.dataset.videoUrl = '';
    img.src = chrome.runtime.getURL('assets/icons/icon48.png');
    button.appendChild(img);
    return button;
  }

  static onClickDownload(event: MouseEvent): void {
    const url = (event.currentTarget as HTMLElement).dataset['videoUrl'];
    url && chrome.runtime.sendMessage({ action: Actions.Download, url });
  }

  static isButtonAttached(videoEl: HTMLElement): boolean {
    return !!videoEl.parentElement
      ?.getElementsByClassName(Page.buttonClassName)
      .length;
  }

  public processAllVideos(): void {
    const allVideoElements = document.getElementsByTagName('video');

    Array.from(allVideoElements).forEach((videoEl) => {
      if (Page.isButtonAttached(videoEl)) {
        return;
      }
      this.attachButton(videoEl);
    });
  }

  public attachButton(videoEl: HTMLVideoElement): void {
    const button = this.buttonElement.cloneNode(true) as HTMLElement;
    button.dataset['videoUrl'] = videoEl.src;
    button.addEventListener('click', Page.onClickDownload);
    videoEl.parentElement?.appendChild(button);
  }

  public onMessage(message: Message): void {
    switch (message.action) {
      case Actions.AddButtons:
        this.processAllVideos();
    }
  }
}

export default Page;
