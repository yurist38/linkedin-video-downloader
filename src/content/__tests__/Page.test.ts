import { mocked } from 'ts-jest/utils';
import { chrome } from '@bumble/jest-chrome';
import { debounce } from 'ts-debounce';
import Page from '../Page';
import { Actions } from '../../types';

jest.mock('ts-debounce');
const debounceMocked = mocked(debounce, false).mockReturnValue(jest.fn());

describe('content/Page', () => {
  let instance: Page;

  beforeAll(() => {
    chrome.runtime.onMessage.addListener = jest.fn();
    MutationObserver.prototype.observe = jest.fn();
  });

  beforeEach(() => {
    instance = new Page();
  });

  it('Properties and constructor are initialized properly', () => {
    expect(Page.buttonClassName).toEqual('linkedin-video-downloader-btn');
    expect(debounceMocked).toHaveBeenCalledTimes(1);
    expect(debounceMocked.mock.calls[0][0]).toEqual(instance.observerCallback);
    expect(debounceMocked.mock.calls[0][1]).toEqual(1000);
    expect(instance.observer).toBeInstanceOf(MutationObserver);
    expect(instance.observer.observe).toHaveBeenCalledTimes(1);
    expect(instance.observer.observe).toHaveBeenCalledWith(window.document, {
      childList: true,
      subtree: true,
    });
    expect(chrome.runtime.onMessage.addListener).toHaveBeenCalledTimes(1);
  });

  it('observerCallback triggers processing all video elements', () => {
    instance.processAllVideos = jest.fn();
    instance.observerCallback();
    expect(instance.processAllVideos).toHaveBeenCalledTimes(1);
  });

  it('Page.createDownloadButton returns a correct element', () => {
    const el = Page.createDownloadButton();
    expect(el).toMatchSnapshot();
  });

  it('Page.onClickDownload is not sending message when url is missing', () => {
    const event = new MouseEvent('test-event');
    Object.defineProperty(event, 'currentTarget', { value: instance.buttonElement });
    Page.onClickDownload(event);
    expect(chrome.runtime.sendMessage).toHaveBeenCalledTimes(0);
  });

  it('Page.onClickDownload is sending message when url is defined', () => {
    const event = new MouseEvent('test-event');
    const url = 'test-url';
    instance.buttonElement.dataset['videoUrl'] = url;
    Object.defineProperty(event, 'currentTarget', { value: instance.buttonElement });
    Page.onClickDownload(event);
    expect(chrome.runtime.sendMessage).toHaveBeenCalledTimes(1);
    expect(chrome.runtime.sendMessage).toHaveBeenCalledWith({ action: Actions.Download, url });
  });

  it('Page.isButtonAttached returns true if the element has the download button already', () => {
    const containerEl = document.createElement('div');
    const videoEl = document.createElement('video');
    const buttonEl = document.createElement('button');
    buttonEl.classList.add(Page.buttonClassName);
    containerEl.appendChild(videoEl);
    containerEl.appendChild(buttonEl);
    expect(Page.isButtonAttached(videoEl)).toEqual(true);
  });

  it('Page.isButtonAttached returns false if the element has no download button', () => {
    const el = document.createElement('div');
    expect(Page.isButtonAttached(el)).toEqual(false);
  });

  it('processAllVideos executes correctly', () => {
    instance.attachButton = jest.fn();
    const el1 = document.createElement('video');
    const el2 = document.createElement('video');
    document.body.appendChild(el1);
    document.body.appendChild(el2);
    Page.isButtonAttached = jest.fn((el) => el === el1);
    instance.processAllVideos();
    expect(instance.attachButton).toHaveBeenCalledTimes(1);
    expect(Page.isButtonAttached).toHaveBeenCalledTimes(2);
  });

  it('attachButton appends the button to the parent element', () => {
    const video = document.createElement('video');
    const container = document.createElement('div');
    container.appendChild(container.appendChild(video));
    instance.attachButton(video);
    expect(container.getElementsByClassName(Page.buttonClassName)).toHaveLength(1);
  });

  it('onMessage reacts on AddButtons action', () => {
    instance.processAllVideos = jest.fn();
    instance.onMessage({ action: Actions.AddButtons });
    expect(instance.processAllVideos).toHaveBeenCalledTimes(1);
  });
});