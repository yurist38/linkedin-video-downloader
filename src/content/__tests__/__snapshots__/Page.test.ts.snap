// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content/Page Page.createDownloadButton returns a correct element 1`] = `
<button
  class="linkedin-video-downloader-btn"
  data-video-url=""
  title="Download this video!"
>
  <img
    src="undefined"
  />
</button>
`;
